teuthology/suite/run.py

```diff
@@ -196,20 +196,55 @@ def choose_ceph_version(self, ceph_hash):
+
+        The branch name value is determined in the following order:
+
+        Use ``--teuthology-branch`` argument value if supplied.
+
+        Use ``TEUTH_BRANCH`` environment variable value if declared.
+
+        If file ``qa/.teuthology_branch`` can be found in the suite repo
+        supplied with ``--suite-repo`` or ``--suite-dir`` and contains
+        non-empty string then use it as the branch name.
+
+        Use ``teuthology_branch`` value if it is set in the one
+        of the teuthology config files ``$HOME/teuthology.yaml``
+        or ``/etc/teuthology.yaml`` correspondingly.
+
+        Use ``master``.
+
+        Generate exception if the branch is not present in the repo.
+        """
         teuthology_branch = self.args.teuthology_branch
-        if teuthology_branch and teuthology_branch != 'master':
-            if not util.git_branch_exists('teuthology', teuthology_branch):
-                exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
-                util.schedule_fail(message=str(exc), name=self.name)
-        elif not teuthology_branch:
+        if teuthology_branch:
+            if  teuthology_branch != 'master':
+                if not util.git_branch_exists('teuthology', teuthology_branch):
+                    exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
+                    util.schedule_fail(message=str(exc), name=self.name)
+        elif 'TEUTH_BRANCH' in os.environ:
+            teuthology_branch = os.environ.get('TEUTH_BRANCH')
+        else:
             # Decide what branch of teuthology to use
-            if util.git_branch_exists('teuthology', self.args.ceph_branch):
-                teuthology_branch = self.args.ceph_branch
-            else:
-                log.info(
-                    "branch {0} not in teuthology.git; will use master for"
-                    " teuthology".format(self.args.ceph_branch))
-                teuthology_branch = 'master'
+            teuthology_branch = None
+            branch_file_path = self.suite_repo_path + '/qa/.teuthology_branch'
+            log.debug('Check file %s exists', branch_file_path)
+            if os.path.exists(branch_file_path):
+                log.debug('Found teuthology branch config file %s',
+                                                        branch_file_path)
+                with open(branch_file_path, 'rt') as f:
+                    teuthology_branch = f.read().strip()
+                    if teuthology_branch:
+                        log.debug(
+                            'The teuthology branch is overridden with %s',
+                                                                teuthology_branch)
+                    else:
+                        log.warning(
+                            'The teuthology branch config is empty, skipping')
+
+            if not teuthology_branch:
+                teuthology_branch = config.get('teuthology_branch', 'master')
```

Comment:
Should the branch be checked for existence here?  It looks like only `--teuthology-branch` command line argument is checked and not `TEUTH_BRANCH`, `.teuthology_branch` or the one from the configuration file.

Reply:
The teuthology-branch is checked later at line 213 (new 248):
```
teuthology_hash = util.git_ls_remote(
             'teuthology',
             teuthology_branch
        )
        if not teuthology_hash:
            exc = BranchNotFoundError(teuthology_branch, build_git_url('teuthology'))
            util.schedule_fail(message=str(exc), name=self.name)
```

Reply:
Then is there a point in checking it separately for `--teuthology-branch` above (lines 223-225)?

Reply:
This is previous behavior and was thinking to delete that code, but I was shy to do it.

Reply:
Dropped the code.